Mathlib/Tactic/MoveAdd.lean

```diff
@@ -0,0 +1,427 @@
+term representing a repeated application of a binary operation and that the summands are the
+last two arguments passed to the operation.
+It returns the expression consisting of the operation with all its arguments already applied,
+except for the last two.
+This is similar to `Lean.Meta.mkAdd, Lean.Meta.mkMul`, except that the resulting operation is
+primed to work with operands of the same type as the ones already appearing in `sum`.
+
+This is useful to rearrange the operands.
+-/
+def prepareOp (sum : Expr) : Expr :=
+let opargs := sum.getAppArgs
+(opargs.toList.take (opargs.size - 2)).foldl (fun x y => Expr.app x y) sum.getAppFn
+
+/--  `sumList op exs` assumes that `op` is the `Name` of a binary operation.
+If `exs` is the list `[e₁, e₂, ..., eₙ]` of `Expr`essions, then `sumList` returns
+`op (op( ... op (op e₁ e₂) e₃) ... eₙ)`.
+-/
+partial
+def sumList1 (prepOp : Expr) : List Expr → Expr
+  | []    => default
+  | [a]    => a
+  | a::as => as.foldl (fun x y => Expr.app (prepOp.app x) y) a
+
+/-- `rankSums op tgt instructions` takes as input
+* the name `op` of a binary operation,
+* an `Expr`ession `tgt`,
+* a list `instructions` of pair `(expression, boolean)`.
+
+It extracts the maximal subexpressions of `tgt` whose head symbol is `op`
+(i.e. the maximal subexpressions that consist only of applications of the binary operation `op`),
+it rearranges the operands of such subexpressions following the order implied by `instructions`
+(as in `reorderUsing`),
+it returns the list of pairs of expressions `(old_sum, new_sum)`, for which `old_sum ≠ new_sum`
+sorted by decreasing value of `Lean.Expr.size`.
+In particular, a subexpression of an `old_sum` can only appear *after* its over-expression.
+-/
+def rankSums (op : Name) (tgt : Expr) (instructions : List (Expr × Bool)) :
+    MetaM (List (Expr × Expr)) := do
+  let sums := ← getOps op (← instantiateMVars tgt)
+  let candidates := ← sums.mapM fun (addends, sum) => do
+    let reord := reorderUsing addends.toList instructions
+    let resummed := sumList1 (prepareOp sum) reord
+    if (resummed != sum) then return some (sum, resummed) else return none
+  return (candidates.toList.reduceOption.toArray.qsort
+    (fun x y : Expr × Expr ↦ (y.1.size  ≤ x.1.size))).toList
+
+/-- `permuteExpr op tgt instructions` takes the same input as `rankSums` and returns the
+expression obtained from `tgt` by replacing all `old_sum`s by the corresponding `new_sum`.
+If there were no required changes, then `permuteExpr` reports this in its second factor. -/
+def permuteExpr (op : Name) (tgt : Expr) (instructions : List (Expr × Bool)) :
+    MetaM Expr := do
+  let permInstructions := ← rankSums op tgt instructions
+  if permInstructions == [] then throwError "The goal is already in the required form"
+  let mut permTgt := tgt
+  for (old, new) in permInstructions do
+    permTgt := permTgt.replace (if · == old then new else none)
+  return permTgt
```

Comment:
`Expr.replace` has a highly optimized implementation, so let's move it to the outer loop: (untested)
```suggestion
  return tgt.replace (fun e => do
    let (_, new) ← permInstructions.find? (fun (old, new) => e == old)
    return new)
```

Reply:
I would expect there is already an implementation of this kind of construction, where we replace a list of expressions with another. But I couldn't find it in the library.

However, if `rankSums` returns a hashmap, we could even write this as `return tgt.replace permInstructions.find?`, so a library function isn't really needed any more.